Zend/zend_weakrefs.c

```diff
@@ -36,14 +36,35 @@ typedef struct _zend_weakmap_iterator {
 } zend_weakmap_iterator;
 
 /* The EG(weakrefs) ht is a map from object address a tagged pointer, that may be one of
- * zend_weakref*, zend_weakmap* or HashTable*. */
+ * zend_weakref*, zend_weakmap* or HashTable*.
+ * ZEND_MM_ALIGNED_OFFSET_LOG2 is at least 2 on supported architectures (pointers to the objects in question are aligned to 4 bytes(1<<2) even on 32-bit systems),
+ * i.e. the least two significant bits of the pointer can be used as a tag. */
 #define ZEND_WEAKREF_TAG_REF 0
 #define ZEND_WEAKREF_TAG_MAP 1
 #define ZEND_WEAKREF_TAG_HT  2
 #define ZEND_WEAKREF_GET_TAG(p) (((uintptr_t) (p)) & 3)
 #define ZEND_WEAKREF_GET_PTR(p) ((void *) (((uintptr_t) (p)) & ~3))
 #define ZEND_WEAKREF_ENCODE(p, t) ((void *) (((uintptr_t) (p)) | (t)))
 
+/* Because php uses the raw numbers as a hash function, raw pointers will lead to hash collisions.
+ * We have a guarantee that the lowest ZEND_MM_ALIGNED_OFFSET_LOG2 bits of a pointer are zero.
+ *
+ * E.g. On most 64-bit platforms, pointers are aligned to 8 bytes, so the least significant 3 bits are always 0 and can be discarded.
+ *
+ * NOTE: This function is only used for EG(weakmaps) and zend_weakmap->ht.
```

Comment:
This seems to be used only for EG(weakrefs), not zend_weakmap->ht. Outdated comment?

We can of course also use it for the weakmap itself, but this will take a bit extra care because the API was recently publicly exported.

Reply:
I'd ran into issues at the time, and incompletely reverted it. I also had overlooked a mass search and replace of obj_addr with obj_key, even where it didn't make sense.

Fixed.